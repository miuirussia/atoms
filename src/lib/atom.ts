// from react types
type SetStateAction<V> = V | ((prev: V) => V);

type Getter = <V>(atom: Atom<V>) => V
type Setter = <V, U, R extends void | Promise<void>>(atom: WritableAtom<V, U, R>, update: U) => R;

type UnmountListener = () => void;
type MountListener<U, R> = (setAtom: (update: U) => R) => UnmountListener | void;

type Read<V> = (get: Getter) => V | Promise<V>;
type Write<U, R extends void | Promise<void> = void> = (get: Getter, set: Setter, update: U) => R;
type WithInitialValue<V> = { init: V };

export interface Atom<V> {
  toString: () => string,
  debugLabel?: string,
  read: Read<V>,
}

export interface WritableAtom<V, U, R extends void | Promise<void> = void> extends Atom<V> {
  write: Write<U, R>,
  onMount?: MountListener<U, R>,
}

export type PrimitiveAtom<V> = WritableAtom<V, SetStateAction<V>>;

export function primitiveAtom<V>(initialValue: V): PrimitiveAtom<V> & WithInitialValue<V> {
  return createAtom(initialValue) as PrimitiveAtom<V> & WithInitialValue<V>;
};

export function atom<V, U, R extends void | Promise<void> = void>(read: Read<V>, write: Write<U, R>): WritableAtom<V, U, R> {
  return createAtom(read, write);
}

export function readOnlyAtom<V>(read: Read<V>): Atom<V> {
  return createAtom(read);
};

export function writeOnlyAtom<U, R extends void | Promise<void>>(write: Write<U, R>): WritableAtom<null, U, R> & WithInitialValue<null> {
  return createAtom(null, write) as WritableAtom<null, U, R> & WithInitialValue<null>;
}

let keyCount: number = 0;

function createAtom<V, U, R extends void | Promise<void>>(read: V | Read<V>, write?: Write<U, R>) {
  const key = `atom${++keyCount}`;

  const spec = {
    toString: () => key,
  } as WritableAtom<V, U, R> & { init?: V };

  if (typeof read === 'function') {
    spec.read = read as Read<V>;
  } else {
    spec.init = read;
    spec.read = get => get(spec);
    spec.write = (get, set, update) => set(spec, typeof update === 'function' ? update(get(spec)) : update);
  }

  if (write) {
    spec.write = write;
  }

  return spec;
}
