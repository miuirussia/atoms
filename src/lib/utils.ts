export const unsafeCoerce = <T>(coercible: any): T => coercible;
