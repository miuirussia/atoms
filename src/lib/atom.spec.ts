import { describe, test, expect, vi } from 'vitest';

import { primitiveAtom, atom, readOnlyAtom, writeOnlyAtom } from './atom';
import { unsafeCoerce } from './utils';

describe('Atom', () => {
  test('primitiveAtom', () => {
    const testAtom = primitiveAtom(0);

    const getter = vi.fn(atom => atom.init);

    expect(testAtom).toMatchSnapshot();
    expect(testAtom.init).toBe(0);

    expect(testAtom.read(getter)).toBe(0);
    expect(getter.mock.calls.length).toBe(1);

    const setter = vi.fn((atom, nextValue) => {
      expect(atom).toBe(testAtom);
      expect(nextValue).toBe(3);
    });
    expect(testAtom.write(getter, unsafeCoerce(setter), 3));
  });
  test('atom', () => {
    const testAtom = atom(async () => 0, async () => {});

    expect(testAtom).toMatchSnapshot();
  });
  test('readOnlyAtom', () => {
    const testAtom = readOnlyAtom(async () => 0);

    expect(testAtom).toMatchSnapshot();
  });
  test('writeOnlyAtom', () => {
    const testAtom = writeOnlyAtom(async () => {});

    expect(testAtom).toMatchSnapshot();
  });
});
