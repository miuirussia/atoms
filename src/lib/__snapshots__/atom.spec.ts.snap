// Vitest Snapshot v1

exports[`Atom > primitiveAtom 1`] = `
{
  "init": 0,
  "read": [Function],
  "toString": [Function],
  "write": [Function],
}
`;
